fdbclient/SpecialKeySpace.actor.cpp

```diff
@@ -115,27 +115,46 @@ ACTOR Future<Void> normalizeKeySelectorActor(SpecialKeySpace* sks, ReadYourWrite
                                              KeyRangeRef boundary, int* actualOffset,
                                              Standalone<RangeResultRef>* result,
                                              Optional<Standalone<RangeResultRef>>* cache) {
+	// If offset < 1, where we need to move left, iter points to the range containing at least one smaller key
+	// (It's a wasting of time to walking through the range whose begin key is same as ks->key)
+	// (rangeContainingKeyBefore itself handles the case where ks->key == Key())
+	// Otherwise, we only need to move right if offset > 1, iter points to the range containing the key
+	// Since boundary.end is always a key in the RangeMap, it is always safe to move right
 	state RangeMap<Key, SpecialKeyRangeBaseImpl*, KeyRangeRef>::Iterator iter =
 	    ks->offset < 1 ? sks->getImpls().rangeContainingKeyBefore(ks->getKey())
 	                   : sks->getImpls().rangeContaining(ks->getKey());
-	while ((ks->offset < 1 && iter->begin() > boundary.begin) || (ks->offset > 1 && iter->begin() < boundary.end)) {
+	while ((ks->offset < 1 && iter->begin() >= boundary.begin) || (ks->offset > 1 && iter->begin() < boundary.end)) {
```

Comment:
The bug is from here.
The code was written under the assumption that the registered range is always not the same as the module range,
so here `=` is not taken into consideration.

Reply:
Thanks for fixing this! I'm having a little difficulty understanding this code because I'm not sure I understand the what this function is doing. Could you add an example to the function comment? What does it mean to normalize a `KeySelector`?

Reply:
There are already some comments above the function, just mention it in case you missed it.

The `key selector` has two attributes `orEqual` and `offset`,
normalize here means to change the key selector to have `orEqual=false` and `offset=1`

For example, if we have a key selector `ks(orEqual=true, offset=-1, key=foo)`
, this means it first finds the largest `key <= foo` and then move left for 1 key.

The standard form with `orEqual=false` and `offset=1`,
means first finding the largest key < given base key and then moving right for 1 key.
Then it just points to the base key.
This standard form is very easy to understand so we want to normalize all selectors to this form.

Reply:
If the key selectors are not in this form, 
we read first to get the right key the selector is pointing to and adjust the key selector to the standard form.
Once we have both begin and end key selectors in this standard form, we are good.